### PR TITLE
Add ability to associate a Service to a Group

### PR DESCRIPTION
**Pull request by montesmoci**

### CODE REVIEW

File: app/src/main/java/app/service/service/ServiceService.java [hunk 1]
```diff
@@ -140,4 +167,62 @@ private boolean serviceDefinitionFormatIsValid(String serviceDefinitionJson) {
...
     }
+
+    public Page<GroupDTO> getPageableGroups(Pageable pageable, String jurisdictionId) {
+        return serviceGroupRepository.findAllByJurisdictionId(jurisdictionId, pageable).map(GroupDTO::new);
+    }
+
+    public GroupDTO createGroup(CreateUpdateGroupDTO requestDTO, String jurisdictionId) {
+        Jurisdiction jurisdiction = jurisdictionRepository.findById(jurisdictionId).get();
+        if (groupAlreadyExists(requestDTO.getName(), jurisdiction)) {
+            return null;
```
Comment (HootDunk) [inline app/src/main/java/app/service/service/ServiceService.java:178]: Can throw exception here

**Review by HootDunk — CHANGES_REQUESTED** [review]

File: app/src/main/java/app/service/service/ServiceService.java [hunk 2]
```diff
@@ -80,6 +88,15 @@ public ServiceDTO createService(CreateServiceDTO serviceDTO, String jurisdiction
                 return null;
             }
         }
+        if (serviceDTO.getGroupId() != null) {
+            Optional<ServiceGroup> serviceGroupOptional = serviceGroupRepository.findByIdAndJurisdiction(serviceDTO.getGroupId(), jurisdiction);
+            if (serviceGroupOptional.isPresent()) {
+                service.setServiceGroup(serviceGroupOptional.get());
+            } else {
+                LOG.error("Service Definition JSON format is invalid.");
+                return null;
```
[the remark this answers is not in the document]
Comment (montesmoci) [reply app/src/main/java/app/service/service/ServiceService.java]: Agreed. I've brought this up in the past and it is planned item for refactor. See [Tech Debt: Error Handling On Front & Backend](https://3.basecamp.com/5573843/buckets/31824829/card_tables/cards/6918499708/edit?focus=kanban_card_title)

File: app/src/main/java/app/JurisdictionAdminController.java [hunk 3]
```diff
@@ -67,9 +70,50 @@ public List<ServiceDTO> updateServiceJson(Long serviceId, @Valid @Body UpdateSer
...
     }
 
+    @Get(uris = { "/groups{?jurisdiction_id}", "/groups.json{?jurisdiction_id}" })
+    @ExecuteOn(TaskExecutors.IO)
+    @RequiresPermissions({LIBRE311_ADMIN_VIEW_SYSTEM, LIBRE311_ADMIN_VIEW_TENANT, LIBRE311_ADMIN_VIEW_SUBTENANT})
+    public HttpResponse<List<GroupDTO>> indexGroups(@Valid Pageable pageable, @Nullable @QueryValue("jurisdiction_id") String jurisdiction_id) {
+        Page<GroupDTO> groupDTOPage = serviceService.getPageableGroups(pageable, jurisdiction_id);
+
+        return HttpResponse.ok(groupDTOPage.getContent())
+                .headers(Map.of(
```
[the remark this answers is not in the document]
Comment (montesmoci) [reply app/src/main/java/app/JurisdictionAdminController.java:80]: Sounds good, I'll make the update. I was following a similar pattern that we've used for list endpoint in root controller See `getServiceRequestsJson` and `indexJson`

File: app/src/main/java/app/JurisdictionAdminController.java [hunk 4]
```diff
@@ -67,9 +70,50 @@ public List<ServiceDTO> updateServiceJson(Long serviceId, @Valid @Body UpdateSer
         return List.of(serviceService.updateService(serviceId, requestDTO, jurisdiction_id));
     }
 
+    @Get(uris = { "/groups{?jurisdiction_id}", "/groups.json{?jurisdiction_id}" })
+    @ExecuteOn(TaskExecutors.IO)
+    @RequiresPermissions({LIBRE311_ADMIN_VIEW_SYSTEM, LIBRE311_ADMIN_VIEW_TENANT, LIBRE311_ADMIN_VIEW_SUBTENANT})
+    public HttpResponse<List<GroupDTO>> indexGroups(@Valid Pageable pageable, @Nullable @QueryValue("jurisdiction_id") String jurisdiction_id) {
```
[the remark this answers is not in the document]
Comment (montesmoci) [reply app/src/main/java/app/JurisdictionAdminController.java:76]: Removing `@Nullable` produces the following error:

```
error: The uri variable [jurisdiction_id] is optional, but the corresponding method argument [java.lang.String jurisdiction_id] is not defined as an Optional or annotated with a Nullable annotation.
    public Page<GroupDTO> indexGroups(@Valid Pageable pageable, @QueryValue("jurisdiction_id") String jurisdiction_id) {

```

[the remark this answers is not in the document]
Comment (HootDunk) [reply app/src/main/java/app/JurisdictionAdminController.java]: Thanks for checking and including the error

**Review by HootDunk — APPROVED** [review]

**Pull request merged**